File: RainLanguage/RainLanguage/Compiler/Context.cpp
﻿#include "Context.h"
#include "AbstractLibrary.h"
#include "../KeyWords.h"
#include "../KernelDeclarations.h"
#include "../Collections/Set.h"

Type MatchBaseType(const String& name)
{
	if(name == KeyWord_bool()) return TYPE_Bool;
	else if(name == KeyWord_byte()) return TYPE_Byte;
	else if(name == KeyWord_char()) return TYPE_Char;
	else if(name == KeyWord_integer()) return TYPE_Integer;
	else if(name == KeyWord_real()) return TYPE_Real;
	else if(name == KeyWord_real2()) return TYPE_Real2;
	else if(name == KeyWord_real3()) return TYPE_Real3;
	else if(name == KeyWord_real4()) return TYPE_Real4;
	else if(name == KeyWord_type()) return TYPE_Type;
	else if(name == KeyWord_string()) return TYPE_String;
	else if(name == KeyWord_handle()) return TYPE_Handle;
	else if(name == KeyWord_entity()) return TYPE_Entity;
	else if(name == KeyWord_array()) return TYPE_Array;
	return Type();
}

Context::Context(const String& source, CompilingSpace* space, List<AbstractSpace*, true>* relies) : source(source), declaration(), compilingSpace(space), relies(relies) {}

Context::Context(const String& source, CompilingDeclaration declaration, CompilingSpace* space, List<AbstractSpace*, true>* relies) : source(source), declaration(declaration), compilingSpace(space), relies(relies) {}

#define CHECK_VISIBILITY \
	if(ContainAny(target.visibility, Visibility::Public | Visibility::Internal)) return true;\
	else if(ContainAny(target.visibility, Visibility::Protected)) return manager->compilingLibrary.GetName(target).source == source;\
	else if(manager->GetDeclaration(target)->space->Contain(compilingSpace->abstract))\
		return ContainAny(target.visibility, Visibility::Space) || manager->compilingLibrary.GetName(target).source == source;\
	return false;

bool Context::IsVisible(DeclarationManager* manager, const CompilingDeclaration& target)
{
	if(target.library == LIBRARY_SELF)
	{
		switch(target.category)
		{
			case DeclarationCategory::Invalid: break;
			case DeclarationCategory::Variable:
			case DeclarationCategory::Function:
			case DeclarationCategory::Enum:
				CHECK_VISIBILITY
			case DeclarationCategory::EnumElement:
				return IsVisible(manager, manager->compilingLibrary.enums[target.definition]->declaration);
			case DeclarationCategory::Struct:
				CHECK_VISIBILITY
			case DeclarationCategory::StructVariable:
				return IsVisible(manager, manager->compilingLibrary.structs[target.definition]->declaration);
			case DeclarationCategory::StructFunction:
			{
				CompilingDeclaration define = manager->compilingLibrary.structs[target.definition]->declaration;
				if(define == declaration) return true;
				return IsVisible(manager, define) && ContainAny(target.visibility, Visibility::Public | Visibility::Internal);
			}
			case DeclarationCategory::Class:
				CHECK_VISIBILITY
			case DeclarationCategory::Constructor:
			case DeclarationCategory::ClassVariable:
			case DeclarationCategory::ClassFunction:
			{
				CompilingDeclaration define = manager->compilingLibrary.classes[target.definition]->declaration;
				if(define == declaration) return true;
				if(IsVisible(manager, define))
					if(ContainAny(target.visibility, Visibility::Public | Visibility::Internal)) return true;
					else if(ContainAny(target.visibility, Visibility::Space)) return manager->GetDeclaration(target)->space->Contain(compilingSpace->abstract);
					else if(declaration.category == DeclarationCategory::Class)
						if(ContainAny(target.visibility, Visibility::Protected))
						{
							Type definType = define.DefineType();
							for(Type index = declaration.DefineType(); index.library != INVALID; index = manager->GetParent(index))
								if(index == definType)
									return true;
						}
			}
			return false;
			case DeclarationCategory::Interface:
				CHECK_VISIBILITY
			case DeclarationCategory::InterfaceFunction:
				return IsVisible(manager, manager->compilingLibrary.interfaces[target.definition]->declaration);
			case DeclarationCategory::Delegate:
			case DeclarationCategory::Task:
			case DeclarationCategory::Native:
				CHECK_VISIBILITY
			case DeclarationCategory::Lambda:
			case DeclarationCategory::LambdaClosureValue:
			case DeclarationCategory::LocalVariable:
			default: break;
		}
	}
	else
	{
		switch(target.category)
		{
			case DeclarationCategory::Invalid: break;
			case DeclarationCategory::Variable:
			case DeclarationCategory::Function:
			case DeclarationCategory::Enum: return ContainAny(target.visibility, Visibility::Public);
			case DeclarationCategory::EnumElement: return ContainAny(manager->GetDeclaration(target)->declaration.visibility, Visibility::Public);
			case DeclarationCategory::Struct: return ContainAny(target.visibility, Visibility::Public);
			case DeclarationCategory::StructVariable: return ContainAny(manager->GetDeclaration(target)->declaration.visibility, Visibility::Public);
			case DeclarationCategory::StructFunction: return ContainAny(manager->GetDeclaration(target)->declaration.visibility, Visibility::Public) && ContainAny(target.visibility, Visibility::Public);
			case DeclarationCategory::Class: return ContainAny(target.visibility, Visibility::Public);
			case DeclarationCategory::Constructor:
			case DeclarationCategory::ClassVariable:
			case DeclarationCategory::ClassFunction:
			{
				CompilingDeclaration define = manager->GetDeclaration(target)->declaration;
				Type definType = define.DefineType();
				if(ContainAny(define.visibility, Visibility::Public))
					if(ContainAny(target.visibility, Visibility::Public))return true;
					else if(declaration.category == DeclarationCategory::Class)
						if(ContainAny(target.visibility, Visibility::Protected))
							for(Type index = declaration.DefineType(); index.library != INVALID; index = manager->GetParent(index))
								if(index == definType)
									return true;
				return false;
			}
			case DeclarationCategory::Interface:  return ContainAny(target.visibility, Visibility::Public);
			case DeclarationCategory::InterfaceFunction: return ContainAny(manager->GetDeclaration(target)->declaration.visibility, Visibility::Public);
			case DeclarationCategory::Delegate:
			case DeclarationCategory::Task:
			case DeclarationCategory::Native: return ContainAny(target.visibility, Visibility::Public);
			case DeclarationCategory::Lambda:
			case DeclarationCategory::LambdaClosureValue:
			case DeclarationCategory::LocalVariable:
			default: break;
		}
	}
	EXCEPTION("无效的定义");
}

bool Context::TryFindSpace(DeclarationManager* manager, const Anchor& name, AbstractSpace*& result)
{
	CompilingSpace* index = compilingSpace;
	while(index)
		if(index->abstract->children.TryGet(name.content, result)) return true;
		else index = index->parent;
	List<AbstractSpace*, true> spaces = List<AbstractSpace*, true>(0);
	for(uint32 i = 0; i < relies->Count(); i++)
		if((*relies)[i]->children.TryGet(name.content, result))
			spaces.Add(result);
	for(uint32 x = 0; x < manager->relies.Count(); x++)
	{
		AbstractLibrary* library = manager->relies[x];
		if(library->name == name.content)
		{
			if(spaces.IndexOf(library) == INVALID) spaces.Add(library);
			break;
		}
	}
	if(spaces.Count())
	{
		result = spaces[0];
		if(spaces.Count() > 1)
		{
			for(uint32 i = 0; i < spaces.Count(); i++)
				MESSAGE3(manager->messages, name, MessageType::ERROR_RELY_SPACE_EQUIVOCAL, spaces[i]->GetFullName(manager->stringAgency));
		}
		return true;
	}
	else if(name.content == manager->kernelLibaray->name)
	{
		result = manager->kernelLibaray;
		return true;
	}
	else return false;
}

void ColletOverrideFunction(Set<CompilingDeclaration, true>& filter, DeclarationManager* manager, AbstractClass* abstractClass, AbstractFunction* function)
{
	AbstractLibrary* abstractLibrary;
	while(abstractClass->parent.library != INVALID)
	{
		abstractLibrary = manager->GetLibrary(abstractClass->parent.library);
		abstractClass = abstractLibrary->classes[abstractClass->parent.index];
		for(uint32 i = 0; i < abstractClass->functions.Count(); i++)
		{
			AbstractFunction* member = abstractLibrary->functions[abstractClass->functions[i]];
			if(member->name == function->name && IsEquals(member->parameters.GetTypes(), 1, function->parameters.GetTypes(), 1))
				filter.Add(member->declaration);
		}
	}
}

bool Context::TryFindDeclaration(DeclarationManager* manager, const Anchor& name, List<CompilingDeclaration, true>& results)
{
	if(declaration.category == DeclarationCategory::Struct)
	{
		AbstractLibrary* library = manager->GetLibrary(declaration.library);
		AbstractStruct* abstractStruct = library->structs[declaration.index];
		for(uint32 i = 0; i < abstractStruct->variables.Count(); i++)
			if(abstractStruct->variables[i]->name == name.content)
			{
				results.Add(abstractStruct->variables[i]->declaration);
				return true;
			}
		for(uint32 i = 0; i < abstractStruct->functions.Count(); i++)
			if(library->functions[abstractStruct->functions[i]]->name == name.content)
				results.Add(library->functions[abstractStruct->functions[i]]->declaration);
		if(results.Count()) return true;
	}
	else if(declaration.category == DeclarationCategory::Class)
	{
		Set<CompilingDeclaration, true> filter(0);
		for(Type index = declaration.DefineType(); index.library != INVALID; index = manager->GetParent(index))
		{
			AbstractLibrary* library = manager->GetLibrary(index.library);
			AbstractClass* abstractClass = library->classes[index.index];
			for(uint32 i = 0; i < abstractClass->variables.Count(); i++)
			{
				AbstractVariable* member = abstractClass->variables[i];
				if(member->name == name.content && IsVisible(manager, member->declaration))
				{
					results.Add(member->declaration);
					return true;//对父类同名字段的覆盖警告放在链接阶段
				}
			}
			for(uint32 i = 0; i < abstractClass->functions.Count(); i++)
			{
				AbstractFunction* member = library->functions[abstractClass->functions[i]];
				if(member->name == name.content && IsVisible(manager, member->declaration) && !filter.Contains(member->declaration))
				{
					results.Add(member->declaration);
					ColletOverrideFunction(filter, manager, abstractClass, member);
				}
			}
		}
		if(results.Count())return true;
	}
	List<CompilingDeclaration, true>* declarations;
	for(AbstractSpace* index = compilingSpace->abstract; index; index = index->parent)
		if(index->declarations.TryGet(name.content, declarations))
		{
			for(uint32 i = 0; i < declarations->Count(); i++)
				if(IsVisible(manager, (*declarations)[i]))
					results.Add((*declarations)[i]);
			if(results.Count() && results.Peek().category != DeclarationCategory::Function && results.Peek().category != DeclarationCategory::Native)
				break;
		}
	if(results.Count() == 1)return true;
	else if(results.Count() > 1)
	{
		for(uint32 x = 0; x < results.Count(); x++)
			if(results[x].category != DeclarationCategory::Function && results[x].category != DeclarationCategory::Native)
			{
				for(uint32 y = 0; y < results.Count(); y++)
					MESSAGE3(manager->messages, name, MessageType::ERROR_DECLARATION_EQUIVOCAL, manager->GetDeclaration(results[y])->GetFullName(manager->stringAgency));
			}
		return true;
	}
	else
	{
		for(uint32 x = 0; x < relies->Count(); x++)
			if((*relies)[x]->declarations.TryGet(name.content, declarations))
				for(uint32 y = 0; y < declarations->Count(); y++)
					if(IsVisible(manager, (*declarations)[y]))
						results.Add((*declarations)[y]);
		if(results.Count() == 1)return true;
		else if(results.Count() > 1)
		{
			for(uint32 x = 0; x < results.Count(); x++)
				if(results[x].category != DeclarationCategory::Function && results[x].category != DeclarationCategory::Native)
				{
					for(uint32 y = 0; y < results.Count(); y++)
						MESSAGE3(manager->messages, name, MessageType::ERROR_DECLARATION_EQUIVOCAL, manager->GetDeclaration(results[y])->GetFullName(manager->stringAgency));
				}
			return true;
		}
	}
	return false;
}

void FindMember(DeclarationManager* manager, const String& name, AbstractInterface* abstractInterface, List<CompilingDeclaration, true>& results)
{
	for(uint32 i = 0; i < abstractInterface->functions.Count(); i++)
	{
		AbstractFunction* member = abstractInterface->functions[i];
		if(member->name == name) results.Add(member->declaration);
	}
	for(uint32 i = 0; i < abstractInterface->inherits.Count(); i++)
	{
		Type& parent = abstractInterface->inherits[i];
		FindMember(manager, name, manager->GetLibrary(parent.library)->interfaces[parent.index], results);
	}
}

bool Context::TryFindMember(DeclarationManager* manager, const String& name, Type type, List<CompilingDeclaration, true>& results)
{
	if(type.dimension) type = TYPE_Array;
	else if(type.code == TypeCode::Enum) type = TYPE_Enum;
	else if(type.code == TypeCode::Delegate) type = TYPE_Delegate;
	else if(type.code == TypeCode::Task) type = TYPE_Task;
	AbstractLibrary* abstractLibrary = manager->GetLibrary(type.library);
	if(type.code == TypeCode::Struct)
	{
		AbstractStruct* abstractStruct = abstractLibrary->structs[type.index];
		for(uint32 i = 0; i < abstractStruct->variables.Count(); i++)
		{
			AbstractVariable* member = abstractStruct->variables[i];
			if(member->name == name)
			{
				results.Add(member->declaration);
				return true;
			}
		}
		for(uint32 i = 0; i < abstractStruct->functions.Count(); i++)
		{
			AbstractFunction* member = abstractLibrary->functions[abstractStruct->functions[i]];
			if(member->name == name && IsVisible(manager, member->declaration))
				results.Add(member->declaration);
		}
	}
	else if(type.code == TypeCode::Handle)
	{
		AbstractClass* abstractClass = abstractLibrary->classes[type.index];
		Set<CompilingDeclaration, true> filter(0);
		while(true)
		{
			for(uint32 i = 0; i < abstractClass->variables.Count(); i++)
			{
				AbstractVariable* member = abstractClass->variables[i];
				if(member->name == name && IsVisible(manager, member->declaration))
				{
					results.Add(member->declaration);
					return true;
				}
			}
			for(uint32 i = 0; i < abstractClass->functions.Count(); i++)
			{
				AbstractFunction* member = abstractLibrary->functions[abstractClass->functions[i]];
				if(member->name == name && IsVisible(manager, member->declaration) && !filter.Contains(member->declaration))
				{
					results.Add(member->declaration);
					ColletOverrideFunction(filter, manager, abstractClass, member);
				}
			}
			if(abstractClass->parent.library == INVALID) break;
			abstractLibrary = manager->GetLibrary(abstractClass->parent.library);
			abstractClass = abstractLibrary->classes[abstractClass->parent.index];
		}
	}
	else if(type.code == TypeCode::Interface) FindMember(manager, name, abstractLibrary->interfaces[type.index], results);
	return results.Count();;
}

void Context::FindDeclaration(DeclarationManager* manager, const List<Anchor>& names, List<CompilingDeclaration, true>& results)
{
	if(names.Count() == 1)
	{
		String name = names[0].content;
		if(declaration.category == DeclarationCategory::Struct)
		{
			AbstractStruct* abstractStruct = manager->selfLibaray->structs[declaration.index];
			for(uint32 x = 0; x < abstractStruct->variables.Count(); x++)
				if(abstractStruct->variables[x]->name == name)
					results.Add(abstractStruct->variables[x]->declaration);
			for(uint32 x = 0; x < abstractStruct->functions.Count(); x++)
				if(manager->selfLibaray->functions[abstractStruct->functions[x]]->name == name)
					results.Add(manager->selfLibaray->functions[abstractStruct->functions[x]]->declaration);
		}
		else if(declaration.category == DeclarationCategory::Class)
		{
			for(Type index = declaration.DefineType(); index.library != INVALID; index = manager->GetParent(index))
			{
				AbstractLibrary* abstractLibrary = manager->GetLibrary(index.library);
				AbstractClass* abstractClass = abstractLibrary->classes[index.index];
				for(uint32 x = 0; x < abstractClass->variables.Count(); x++)
				{
					AbstractVariable* member = abstractClass->variables[x];
					if(member->name == name && IsVisible(manager, member->declaration))
						results.Add(member->declaration);
				}
				for(uint32 x = 0; x < abstractClass->functions.Count(); x++)
				{
					AbstractFunction* member = abstractLibrary->functions[abstractClass->functions[x]];
					if(member->name == name && IsVisible(manager, member->declaration))
						results.Add(member->declaration);
				}
			}
		}
		if(results.Count())return;
		List<CompilingDeclaration, true>* declarations;
		for(AbstractSpace* index = compilingSpace->abstract; index; index = index->parent)
			if(index->declarations.TryGet(name, declarations))
			{
				results.Add(*declarations);
				return;
			}
		for(uint32 x = 0; x < relies->Count(); x++)
		{
			AbstractSpace* space = (*relies)[x];
			if(space->declarations.TryGet(name, declarations))
				results.Add(*declarations);
		}
		if(results.Count()) return;
		Type type = MatchBaseType(name);
		if(type.IsValid())
			results.Add(manager->GetDeclaration(type)->declaration);
	}
	else if(names.Count() > 1)
	{
		AbstractSpace* space;
		if(TryFindSpace(manager, names[0], space))
		{
			for(uint32 i = 1; i < names.Count() - 1; i++)
				if(!space->children.TryGet(names[i].content, space))
					return;
			List<CompilingDeclaration, true>* declarations;
			if(space->declarations.TryGet(names.Peek().content, declarations))
			{
				results.Add(*declarations);
				return;
			}
		}
	}
}

void Context::FindOperators(DeclarationManager* manager, const String& name, List<CompilingDeclaration, true>& results)
{
	List<CompilingDeclaration, true>* declarations;
	if(manager->kernelLibaray->declarations.TryGet(name, declarations))
		results.Add(*declarations);
	for(AbstractSpace* index = compilingSpace->abstract; index; index = index->parent)
		if(index->declarations.TryGet(name, declarations))
			results.Add(*declarations);
	for(uint32 x = 0; x < relies->Count(); x++)
	{
		AbstractSpace* space = (*relies)[x];
		if(space != manager->kernelLibaray && space->declarations.TryGet(name, declarations))
			results.Add(*declarations);
	}
#ifdef DEBUG
	for(uint32 i = 0; i < results.Count(); i++)
		if(results[i].category != DeclarationCategory::Function)
			EXCEPTION("操作类型必须是全局函数");
#endif // DEBUG
	}

File: RainLanguageServer/RainLanguage/Context.cs
using System.Text;

namespace RainLanguageServer.RainLanguage
{
    internal class Context(TextDocument document, CompilingSpace space, HashSet<CompilingSpace> relies, CompilingDeclaration? declaration)
    {
        public readonly TextDocument document = document;
        public readonly CompilingSpace space = space;
        public readonly HashSet<CompilingSpace> relies = relies;
        public readonly CompilingDeclaration? declaration = declaration;

        private bool IsVisiable(ASTManager manager, Declaration declaration, bool isMember)
        {
            if (isMember)
            {
                if (IsVisiable(manager, manager.GetDeclaringDeclaration(declaration)!.declaration, false))
                    if (declaration.visibility.ContainAny(Visibility.Public | Visibility.Internal)) return true;
                return false;
            }
            else
            {
                if (declaration.visibility.ContainAny(Visibility.Public | Visibility.Internal)) return true;
                else
                {
                    var compiling = manager.GetDeclaration(declaration)!;
                    if (compiling.space.Contain(space))
                    {
                        if (declaration.visibility.ContainAny(Visibility.Space)) return true;
                        else return compiling.name.start.document == document;
                    }
                    return false;
                }
            }
        }
        public bool IsVisiable(ASTManager manager, Declaration declaration)
        {
            if (declaration.library == space.Library.name)
            {
                switch (declaration.category)
                {
                    case DeclarationCategory.Invalid: break;
                    case DeclarationCategory.Variable:
                    case DeclarationCategory.Function:
                    case DeclarationCategory.Enum:
                        return IsVisiable(manager, declaration, false);
                    case DeclarationCategory.EnumElement:
                        return IsVisiable(manager, declaration, true);
                    case DeclarationCategory.Struct:
                        return IsVisiable(manager, declaration, false);
                    case DeclarationCategory.StructVariable:
                        return IsVisiable(manager, declaration, true);
                    case DeclarationCategory.StructFunction:
                        {
                            var define = manager.GetDeclaringDeclaration(declaration)!;
                            if (define == this.declaration) return true;
                            if (IsVisiable(manager, define.declaration, false))
                                if (declaration.visibility.ContainAny(Visibility.Public | Visibility.Internal)) return true;
                                else if (declaration.visibility.ContainAny(Visibility.Space)) return define.space.Contain(space);
                        }
                        return false;
                    case DeclarationCategory.Class:
                        return IsVisiable(manager, declaration, false);
                    case DeclarationCategory.Constructor:
                    case DeclarationCategory.ClassVariable:
                    case DeclarationCategory.ClassFunction:
                        {
                            var define = manager.GetDeclaringDeclaration(declaration)!;
                            if (define == this.declaration) return true;
                            if (IsVisiable(manager, define.declaration, false))
                                if (declaration.visibility.ContainAny(Visibility.Public | Visibility.Internal)) return true;
                                else if (declaration.visibility.ContainAny(Visibility.Space)) return define.space.Contain(space);
                                else if (this.declaration?.declaration.category == DeclarationCategory.Class)
                                    if (declaration.visibility.ContainAny(Visibility.Protected))
                                    {
                                        var defineType = define.declaration.GetDefineType();
                                        foreach (var index in manager.GetInheritIterator(this.declaration.declaration.GetDefineType()))
                                            if (index == defineType) return true;
                                    }
                        }
                        return false;
                    case DeclarationCategory.Interface:
                        return IsVisiable(manager, declaration, false);
                    case DeclarationCategory.InterfaceFunction:
                        return IsVisiable(manager, declaration, true);
                    case DeclarationCategory.Delegate:
                    case DeclarationCategory.Task:
                    case DeclarationCategory.Native:
                        return IsVisiable(manager, declaration, false);
                }
            }
            else
            {
                switch (declaration.category)
                {
                    case DeclarationCategory.Invalid: break;
                    case DeclarationCategory.Variable:
                    case DeclarationCategory.Function:
                    case DeclarationCategory.Enum: return declaration.visibility.ContainAny(Visibility.Public);
                    case DeclarationCategory.EnumElement: return manager.GetDeclaringDeclaration(declaration)!.declaration.visibility.ContainAny(Visibility.Public);
                    case DeclarationCategory.Struct: return declaration.visibility.ContainAny(Visibility.Public);
                    case DeclarationCategory.StructVariable: return manager.GetDeclaringDeclaration(declaration)!.declaration.visibility.ContainAny(Visibility.Public);
                    case DeclarationCategory.StructFunction: return manager.GetDeclaringDeclaration(declaration)!.declaration.visibility.ContainAny(Visibility.Public) && declaration.visibility.ContainAny(Visibility.Public);
                    case DeclarationCategory.Class: return declaration.visibility.ContainAny(Visibility.Public);
                    case DeclarationCategory.Constructor:
                    case DeclarationCategory.ClassVariable:
                    case DeclarationCategory.ClassFunction:
                        {
                            var define = manager.GetDeclaringDeclaration(declaration);
                            if (define!.declaration.visibility.ContainAny(Visibility.Public))
                                if (declaration.visibility.ContainAny(Visibility.Public)) return true;
                                else if (this.declaration != null && this.declaration.declaration.category == DeclarationCategory.Class && declaration.visibility.ContainAny(Visibility.Protected))
                                    for (var index = this.declaration as CompilingClass; index != null; index = manager.GetSourceDeclaration(index.parent) as CompilingClass)
                                        if (index == define)
                                            return true;
                        }
                        break;
                    case DeclarationCategory.Interface: return declaration.visibility.ContainAny(Visibility.Public);
                    case DeclarationCategory.InterfaceFunction: return manager.GetDeclaringDeclaration(declaration)!.declaration.visibility.ContainAny(Visibility.Public);
                    case DeclarationCategory.Delegate:
                    case DeclarationCategory.Task:
                    case DeclarationCategory.Native: return declaration.visibility.ContainAny(Visibility.Public);
                }
            }
            return false;
        }
        public bool TryFindSpace(ASTManager manager, TextRange name, out CompilingSpace? space, MessageCollector collector)
        {
            var targetName = name.ToString();
            for (var index = this.space; index != null; index = index.parent)
                if (index.children.TryGetValue(targetName, out space)) return true;
            var results = new HashSet<CompilingSpace>();
            foreach (var rely in relies)
                if (rely.children.TryGetValue(targetName, out space)) results.Add(space!);
            if (manager.relies.TryGetValue(targetName, out var library))
                results.Add(library);
            if (results.Count > 0)
            {
                if (results.Count > 1)
                {
                    var message = new StringBuilder().AppendLine("依赖空间不明确");
                    foreach (var result in results)
                        message.AppendLine(result.GetFullName());
                    collector.Add(name, CErrorLevel.Error, message.ToString());
                }
                space = results.First();
                return true;
            }
            else if (targetName == manager.kernel.name)
            {
                space = manager.kernel;
                return true;
            }
            space = default;
            return false;
        }
        public bool TryFindDeclaration(ASTManager manager, TextRange name, out List<CompilingDeclaration> results, MessageCollector collector)
        {
            results = [];
            var targetName = name.ToString();
            if (declaration?.declaration.category == DeclarationCategory.Struct)
            {
                var compilingStruct = (CompilingStruct)declaration;
                foreach (var variable in compilingStruct.variables)
                    if (variable.name == targetName)
                    {
                        results.Add(variable);
                        return true;
                    }
                foreach (var function in compilingStruct.functions)
                    if (function.name == targetName)
                        results.Add(function);
                if (results.Count > 0) return true;
            }
            else if (declaration?.declaration.category == DeclarationCategory.Class)
            {
                var filter = new HashSet<CompilingDeclaration>();
                foreach (var index in manager.GetInheritIterator(declaration.declaration.GetDefineType()))
                {
                    var @class = (CompilingClass)manager.GetSourceDeclaration(index)!;
                    foreach (var variable in @class.variables)
                        if (variable.name == targetName && IsVisiable(manager, variable.declaration))
                        {
                            results.Add(variable);
                            return true;
                        }
                    foreach (var function in @class.functions)
                        if (function.name == targetName && IsVisiable(manager, function.declaration) && !filter.Contains(function))
                        {
                            results.Add(function);
                            CollectOverideFunctions(filter, manager, @class, function);
                        }
                }
                if (results.Count > 0) return true;
            }

            for (var index = space; index != null; index = index.parent)
                if (index.declarations.TryGetValue(targetName, out var declarations))
                {
                    foreach (var item in declarations)
                        if(IsVisiable(manager, item.declaration))
                            results.Add(item);
                    if (results.Count > 0 && results[0].declaration.category != DeclarationCategory.Function && results[0].declaration.category != DeclarationCategory.Native) break;
                }
            if (results.Count == 1) return true;
            else if (results.Count > 1)
            {
                foreach (var declaration in results)
                    if (declaration.declaration.category != DeclarationCategory.Function && declaration.declaration.category != DeclarationCategory.Native)
                    {
                        var builder = new StringBuilder().AppendLine("查找申明不明确");
                        foreach (var item in results)
                            builder.AppendLine(item.GetFullName());
                        collector.Add(name, CErrorLevel.Error, builder.ToString());
                        break;
                    }
                return true;
            }
            else
            {
                foreach (var rely in relies)
                    if (rely.declarations.TryGetValue(targetName, out var declarations))
                        foreach (var declaration in declarations)
                            if (IsVisiable(manager, declaration.declaration))
                                results.Add(declaration);
                if (results.Count == 1) return true;
                else if (results.Count > 1)
                {
                    foreach (var declaration in results)
                        if (declaration.declaration.category != DeclarationCategory.Function && declaration.declaration.category != DeclarationCategory.Native)
                        {
                            var builder = new StringBuilder().AppendLine("查找申明不明确");
                            foreach (var item in results)
                                builder.AppendLine(item.GetFullName());
                            collector.Add(name, CErrorLevel.Error, builder.ToString());
                            break;
                        }
                    return true;
                }
            }
            return false;
        }
        private static void CollectOverideFunctions(HashSet<CompilingDeclaration> filter, ASTManager manager, CompilingClass @class, CompilingFunction function)
        {
            var name = function.name.ToString();
            foreach (var index in manager.GetInheritIterator(manager.GetParent(@class.declaration.GetDefineType())))
                if (manager.GetSourceDeclaration(index) is CompilingClass compiling)
                    foreach (var item in compiling.functions)
                        if (item.name == name && item.declaration.signature == function.declaration.signature)
                            filter.Add(item);
        }
        private static void CollectFunctions(ASTManager manager, string name, CompilingInterface @interface, List<CompilingDeclaration> results)
        {
            foreach (var item in @interface.callables)
                if (item.name == name)
                    results.Add(item);
            foreach (var item in @interface.inherits)
                CollectFunctions(manager, name, (CompilingInterface)manager.GetSourceDeclaration(item)!, results);
        }
        public bool TryFindMember(ASTManager manager, TextRange name, Type type, out List<CompilingDeclaration> results)
        {
            results = [];
            if (type.dimension > 0) type = Type.ARRAY;
            else if (type.code == TypeCode.Enum) type = Type.ENUM;
            else if (type.code == TypeCode.Task) type = Type.TASK;
            else if (type.code == TypeCode.Delegate) type = Type.DELEGATE;
            var targetName = name.ToString();
            var declaration = manager.GetSourceDeclaration(type)!;
            if (type.code == TypeCode.Struct)
            {
                var compiling = (CompilingStruct)declaration;
                foreach (var item in compiling.variables)
                    if (item.name == targetName)
                    {
                        results.Add(item);
                        return true;
                    }
                foreach (var item in compiling.functions)
                    if (item.name == targetName && IsVisiable(manager, item.declaration))
                        results.Add(item);
            }
            else if (type.code == TypeCode.Handle)
            {
                var filter = new HashSet<CompilingDeclaration>();
                foreach (var index in manager.GetInheritIterator(declaration.declaration.GetDefineType()))
                    if (manager.GetSourceDeclaration(index) is CompilingClass compiling)
                    {
                        foreach (var item in compiling.variables)
                            if (item.name == targetName && IsVisiable(manager, item.declaration))
                            {
                                results.Add(item);
                                return true;
                            }
                        foreach (var item in compiling.functions)
                            if (item.name == targetName && IsVisiable(manager, item.declaration) && !filter.Contains(item))
                            {
                                results.Add(item);
                                CollectOverideFunctions(filter, manager, compiling, item);
                            }
                    }
            }
            else if (type.code == TypeCode.Interface) CollectFunctions(manager, targetName, (CompilingInterface)declaration, results);
            return results.Count > 0;
        }
        public bool TryFindDeclaration(ASTManager manager, List<TextRange> name, out List<CompilingDeclaration> results, MessageCollector collector)
        {
            results = [];
            if (name.Count == 1)
            {
                var type = LexicalTypeExtend.Parse(name[0].ToString());
                switch (type)
                {
                    case LexicalType.KeyWord_bool: results.Add(manager.GetSourceDeclaration(Type.BOOL)!); break;
                    case LexicalType.KeyWord_byte: results.Add(manager.GetSourceDeclaration(Type.BYTE)!); break;
                    case LexicalType.KeyWord_char: results.Add(manager.GetSourceDeclaration(Type.CHAR)!); break;
                    case LexicalType.KeyWord_integer: results.Add(manager.GetSourceDeclaration(Type.INT)!); break;
                    case LexicalType.KeyWord_real: results.Add(manager.GetSourceDeclaration(Type.REAL)!); break;
                    case LexicalType.KeyWord_real2: results.Add(manager.GetSourceDeclaration(Type.REAL2)!); break;
                    case LexicalType.KeyWord_real3: results.Add(manager.GetSourceDeclaration(Type.REAL3)!); break;
                    case LexicalType.KeyWord_real4: results.Add(manager.GetSourceDeclaration(Type.REAL4)!); break;
                    case LexicalType.KeyWord_type: results.Add(manager.GetSourceDeclaration(Type.TYPE)!); break;
                    case LexicalType.KeyWord_string: results.Add(manager.GetSourceDeclaration(Type.STRING)!); break;
                    case LexicalType.KeyWord_entity: results.Add(manager.GetSourceDeclaration(Type.ENTITY)!); break;
                    case LexicalType.KeyWord_handle: results.Add(manager.GetSourceDeclaration(Type.HANDLE)!); break;
                    case LexicalType.KeyWord_interface: results.Add(manager.GetSourceDeclaration(Type.INTERFACE)!); break;
                    case LexicalType.KeyWord_delegate: results.Add(manager.GetSourceDeclaration(Type.DELEGATE)!); break;
                    case LexicalType.KeyWord_task: results.Add(manager.GetSourceDeclaration(Type.TASK)!); break;
                    case LexicalType.KeyWord_array: results.Add(manager.GetSourceDeclaration(Type.ARRAY)!); break;
                    default:
                        var targetName = name[0].ToString();
                        if (declaration != null)
                        {
                            if (declaration.declaration.category == DeclarationCategory.Struct)
                            {
                                var compiling = (CompilingStruct)declaration;
                                foreach (var item in compiling.variables)
                                    if (item.name == targetName)
                                        results.Add(item);
                                foreach (var item in compiling.functions)
                                    if (item.name == targetName)
                                        results.Add(item);
                            }
                            else if (declaration.declaration.category == DeclarationCategory.Class)
                            {
                                for (var index = (CompilingClass)declaration; index != null; index = manager.GetSourceDeclaration(index.parent) as CompilingClass)
                                {
                                    foreach (var item in index.variables)
                                        if (item.name == targetName && IsVisiable(manager, item.declaration))
                                            results.Add(item);
                                    foreach (var item in index.functions)
                                        if (item.name == targetName && IsVisiable(manager, item.declaration))
                                            results.Add(item);
                                }
                            }
                            if (results.Count > 0) return true;
                        }
                        for (var index = space; index != null; index = index.parent)
                            if (index.declarations.TryGetValue(targetName, out var declarations))
                            {
                                results.AddRange(declarations);
                                return true;
                            }
                        foreach (var rely in relies)
                            if (rely.declarations.TryGetValue(targetName, out var declarations))
                                results.AddRange(declarations);
                        break;
                }
            }
            else if (name.Count > 1)
            {
                if (TryFindSpace(manager, name[0], out var space, collector))
                {
                    for (var i = 1; i < name.Count - 1; i++)
                        if (!space!.children.TryGetValue(name[i].ToString(), out space))
                            return false;
                    if (space!.declarations.TryGetValue(name[^1].ToString(), out results!)) return true;
                    else results = [];
                }
            }
            return results.Count > 0;
        }
        public List<CompilingDeclaration> FindOperator(ASTManager manager, string name)
        {
            var result = new List<CompilingDeclaration>();
            if (manager.kernel.declarations.TryGetValue(name, out var declarations))
                result.AddRange(declarations);
            for (CompilingSpace? index = space; index != null; index = index.parent)
                if (index.declarations.TryGetValue(name, out declarations))
                    result.AddRange(declarations);
            foreach (var rely in relies)
                if (rely.declarations.TryGetValue(name, out declarations))
                    result.AddRange(declarations);
            return result;
        }
    }
}
